src/script_opt/UsageAnalyzer.cc

```diff
@@ -0,0 +1,224 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+
+#include "zeek/script_opt/UsageAnalyzer.h"
+
+#include "zeek/module_util.h"
+#include "zeek/script_opt/IDOptInfo.h"
+
+namespace zeek::detail
+	{
+
+// The names of identifiers that correspond to events not-previously-known
+// before their declaration in the scripts.
+std::unordered_set<std::string> script_events;
+
+void register_new_event(const IDPtr& id)
+	{
+	script_events.insert(id->Name());
+	}
+
+UsageAnalyzer::UsageAnalyzer(std::vector<FuncInfo>& funcs)
+	{
+	// Setting a scope cues ID::Traverse to delve into function values.
+	current_scope = global_scope();
+
+	FindSeeds(reachables);
+	FullyExpandReachables();
+
+	// At this point, we've done the complete reachability analysis.
+	// Report out on unreachables.  We do this in two steps: first,
+	// unreachable events/hooks, and then unreachable functions.  We
+	// split the two because we don't want to ding a function as being
+	// unreachable if there's an (unreachable) event-or-hook that calls
+	// it, since presumably the real problem is the latter being an
+	// orphan, rather than the function.
+
+	auto& globals = global_scope()->Vars();
+
+	for ( auto& gpair : globals )
+		{
+		auto id = gpair.second.get();
+		auto& t = id->GetType();
+
+		if ( t->Tag() != TYPE_FUNC )
+			continue;
+
+		if ( t->AsFuncType()->Flavor() == FUNC_FLAVOR_FUNCTION )
+			continue;
+
+		if ( reachables.count(id) > 0 )
+			continue;
+
+		auto flavor = t->AsFuncType()->FlavorString();
+		auto loc = id->GetLocationInfo();
+
+		reporter->Warning("%s %s (%s:%d): cannot be invoked", flavor.c_str(), id->Name(),
```

Comment:
When I first encountered this, the emitted warning wasn't clear to me. Maybe this could be stated in a way which formulates what a user be do about this and maybe also more clearly call out why we think this is an issue. It would also be nice if the location format of this diagnostic would follow what Zeek already does for source parse errors so that downstream tooling (e.g., editors) can use a single error format to parse out location and error.

Maybe something like:

> warning in /tmp/bla.zeek, line 2: found event handle for non-existing event "does_not_exist" which will never be invoked

Reply:
Thanks for the thought.  I've gone ahead and regularized these